Sources/SourceKitLSP/Swift/CodeActions/PackageManifestEdits.swift

```diff
@@ -0,0 +1,243 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import PackageModel
+import PackageModelSyntax
+import SwiftRefactor
+import SwiftSyntax
+
+/// Syntactic code action provider to provide refactoring actions that
+/// edit a package manifest.
+struct PackageManifestEdits: SyntaxCodeActionProvider {
+  static func codeActions(in scope: SyntaxCodeActionScope) -> [CodeAction] {
+    guard let token = scope.firstToken,
+      let call = token.findEnclosingCall()
+    else {
+      return []
+    }
+
+    return addTestTargetActions(call: call, in: scope) + addProductActions(call: call, in: scope)
+  }
+
+  /// Produce code actions to add test target(s) if we are currently on
+  /// a target for which we know how to create a test.
+  static func addTestTargetActions(
+    call: FunctionCallExprSyntax,
+    in scope: SyntaxCodeActionScope
+  ) -> [CodeAction] {
+    guard let calledMember = call.findMemberAccessCallee(),
+      targetsThatAllowTests.contains(calledMember),
+      let targetName = call.findStringArgument(label: "name")
+    else {
+      return []
+    }
+
+    do {
+      // Describe the target we are going to create.
+      let target = try TargetDescription(
+        name: "\(targetName)Tests",
+        dependencies: [.byName(name: targetName, condition: nil)],
+        type: .test
+      )
+
+      let edits = try AddTarget.addTarget(target, to: scope.file)
+      return [
+        CodeAction(
+          title: "Add test target",
+          kind: .refactor,
+          edit: edits.asWorkspaceEdit(snapshot: scope.snapshot)
+        )
+      ]
+    } catch {
+      return []
+    }
+  }
+
+  /// A list of target kinds that allow the creation of tests.
+  static let targetsThatAllowTests: Set<String> = [
```

Comment:
I know this isn't particularly performance-sensitive, but this should probably be an Array. A Set is overkill.